### PR TITLE
Add support for multi-query and grouped-query attention

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR adds support for multi-query attention (MQA) and grouped-query attention (GQA). Specifically, it enables the use of MQA and GQA in FlashAttention() and UnfusedDotProductAttention. MQA and MHA are special cases of GQA, i.e. when `num_gqa_groups`=`1` and `num_heads`. GQA only applies to keys and values, not queries.

### CODE REVIEW

File: tests/pytorch/test_fused_attn.py [hunk 1]
```diff
@@ -208,6 +213,105 @@ def _run_transformer_layer(dtype, bs, config, backend):
...
+
+    num_querys_per_gqa_group = find_factors(config.num_attention_heads)
+
+    for num_q_per_gqa_group in num_querys_per_gqa_group:
+        flash_attn_fwd, flash_attn_bwd = _run_transformer_layer_gqa(
+                dtype, bs, config, "FlashAttention", num_q_per_gqa_group)
+        unfused_attn_fwd, unfused_attn_bwd = _run_transformer_layer_gqa(
+                dtype, bs, config, "UnfusedDotProductAttention", num_q_per_gqa_group)
+
+        atol, rtol = (5e-1, 5e-1) if dtype == torch.bfloat16 else (5e-1, 5e-1)
```
Comment (ksivaman) [inline tests/pytorch/test_fused_attn.py:240]:
```suggestion
        atol, rtol = 5e-1, 5e-1
```

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -718,6 +727,14 @@ class DotProductAttention(torch.nn.Module):
                          number of attention heads in the transformer layer.
     kv_channels : int
                 number of key-value channels.
+    num_gqa_groups : Optional[int] = None
+                    number of GQA groups in the transformer layer.
+                    Grouped Query Attention is described in
+                    `this paper <https://arxiv.org/pdf/2305.13245.pdf>`_.
+                    This only affects the keys and values, not the querys.
```
Comment (ksivaman) [inline transformer_engine/pytorch/attention.py:734]:
```suggestion
                    This only affects the keys and values, not the queries.
```

**Comment by blahBlahhhJ** [comment]
Hi, overall LGTM with only one small thing. I see @ksivaman mentioned that this PR uses a PR in MegatronLM as a reference. The main difference I see is that in TE you use a Q projection and a KV projection while in megatron it's packed to be a single QKV projection matrix. Is there any particular reasons why you separate them? I think it would be better if the implementations are consistent so that people don't need to convert checkpoints back and forth. especially because I see megatron core uses some TE layers internally.

**Comment by cyanguwa** [comment]
Hi @blahBlahhhJ ,

With MQA and GQA, the kv projection size might be different from the query projection size. That itself isn't a problem for the projection gemm, but when we do ColumnParallel and the split of parameters (into q, k, v, in TE), the `q_projection_size + 2 * kv_projection_size` might not divide well by 3. To be safe and to make the TP balancing easier (I'm sure we can parallelize it too with a non-balanced TP situation, if we really want to do it), I coded the projection as Q and KV rather than QKV.

**Review by ksivaman — APPROVED** [review]
LGTM

**Pull request merged**